cli/src/torii/routing.rs

```diff
@@ -135,49 +73,34 @@ pub(crate) async fn handle_queries(
     pagination: Pagination,
     sorting: Sorting,
     request: VersionedSignedQuery,
-) -> Result<Scale<VersionedPaginatedQueryResult>> {
-    let VersionedSignedQuery::V1(request) = request;
-    let request: VerifiedQuery = request.try_into().map_err(ValidationFail::from)?;
-
-    let (result, filter) = {
-        let mut wsv = sumeragi.wsv_clone();
-        let (valid_request, filter) = request.validate(&mut wsv)?;
-        let original_result = valid_request.execute(&wsv).map_err(ValidationFail::from)?;
-        (filter.filter(original_result), filter)
-    };
+) -> Result<Scale<VersionedQueryResult>> {
+    let mut wsv = sumeragi.wsv_clone();
```

Comment:
Need to find a way to elide this clone. 

From a common-sense point of view, queries should **not** mutate the `wsv` and as such we not only don't need to clone it, we only need an immutable borrow. If we could externalise the mutation, segregate it and make it reversible, the `&wsv` should be enough.

Reply:
yes and no. Cloning is not here because of mutation but to relieve contention on re-entrant pagination. We don't want to clone on queries that return a non-iterable value, but for paginated queries we have to store the snapshot of the wsv

Reply:
You're right. I need to reformulate the comment. 

What threw me off is that this logic is not supposed to be in the `routing` module of Torii. 

## Cloning WSV

It's fine if the clones are cheap and that's what we intend to have eventually. In that case this logic should be intrinsic to how all queries are handled. In other words, the fact that we're cloning isn't an interface detail anymore. We should describe whether a query needs to clone the `wsv` in its `ValidQuery` impl,  if we don't clone in all circumstances. 

Unfortunately, we also can't make the persistent data structure for a cheap WSV with off-the-shelf components. We need to go custom and this is why for all intents and purposes until `wsv` clones are cheap we should treat them as expensive. 

## COW WSV

Given expensive `wsv` clones, and/or expensive `clone()` + `execute()` combined operations, it might make sense to hook into the event system and ensure that the `wsv` is only updated if a relevant event was emitted somewhere else. This gives you a more granular `Cow`. 

Another point is that evaluating the query to completion might be as fast as cloning the `wsv`. In that case we should be even more granular about what we clone and what we borrow.

Reply:
## Querying by block height

In either case, it might also be worth considering persistent `wsv` snapshots for each block height. Each query optionally includes the block height for which it must execute. If an appropriate `wsv` is stored in the cache, it goes as before. The `wsv` cloning can be done asynchronously so we don't block normal operations and lazily so we don't need to clone each time we receive a query. 

Once the snapshot is stored, the data there is immutable, so multiple queries can operate on the same data simultaneously. This would insert a delay until the current block is committed, but until that happens the information isn't really *in* the blockchain. 

The logic can then be reused for a lightweight archival node.

Reply:
> Given expensive wsv clones, and/or expensive clone() + execute() combined operations, it might make sense to hook into the event system and ensure that the wsv is only updated if a relevant event was emitted somewhere else. This gives you a more granular Cow.

We already do that, we can watch for updates of `wsv` through sumeragi `wsv_updated` method. Problem is that it isn't easily integrated into `torii`.

Reply:
> We should describe whether a query needs to clone the wsv in its ValidQuery impl, if we don't clone in all circumstances.

I think this insight alone will solve a lot of headache. I can do this if `ValidQuery::execute(&wsv)` will never return references to the `wsv` and if it returns iterator then `Iterator::Item` will never contain a reference to the `wsv`. Considering how `ValidQuery::execute` is used I think this approach should work.

Reply:
> We already do that, we can watch for updates of wsv through sumeragi wsv_updated method. Problem is that it isn't easily integrated into torii.

Hence the very first thing I mentioned was that this code **does not belong in `torii`**.

Reply:
I think I resolved most concerns related to this